tm_sympy - Correctly handle exceptions and output formatting

mardukbp opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->


#### Brief description of what is fixed or changed

Given some input code, tm_sympy checks whether it is an expression or a statement. Executing a statement does not generate output. If the result of evaluating an expression is a SymPy object, tm_sympy outputs its LaTeX representation.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* other
  * [TeXmacs] Exceptions are correctly handled and the interactive session does not hang anymore.
  * [TeXmacs] SymPy output is sent to TeXmacs as LaTeX. Any other output is displayed verbatim.
<!-- END RELEASE NOTES -->

mardukbp committed:
tm_sympy - Improved output

Depending on the type of the output it is processed accordingly and sent to TeXmacs.

asmeurer commented:
I'm still not clear how to actually use the plugin in TeXmacs. 

Also I noticed that TeXmacs lists several plugins for other CASes in its help menu. Would it make sense to move the TeXmacs plugin from the SymPy repo to the TeXmacs repo, so that they can maintain it? As far as I know, no SymPy maintainers use the plugin, so it isn't being very well maintained here.

mardukbp committed:
Correctly handle exceptions and output format

Given some input code, check whether it is an expression or a statement. Executing a statement does not generate output.
The result of evaluating a Python expression can be of two types: either it has a LaTeX representation or it does not.

mardukbp committed:
Merge branch 'master' into improved-output

asmeurer commented:
OK, now that I know how to use the plugin, I can't get it to actually install. My ~/.TeXmacs/plugins/sympy/ is a copy of data/TeXmacs. Should we include a small install script with the plugin? 

> I would like to suggest that SymPy advertises the TeXmacs interface.

Having a README included with the TeXmacs plugin a good start for this. Any other places you see where we should advertise it? 

Also, one of the reasons I want to include the plugin with TeXmacs itself is so that TeXmacs advertises SymPy in its list of builtin plugins. It already includes several other CASes, both open and closed.

asmeurer commented:
I am still unable to test this. When I start TeXmacs and try to open a menu it just hangs.

mardukbp commented:
> It eventually stopped hanging and I was able to start the plugin.
Which version of macOS do you have? Now I am using 10.11 and have not experienced any issues.
 
> In the terminal it says
> TeXmacs] Error: cannot start 'tm_sympy --texmacs'

I believe `chmod +x tm_sympy` should fix this.

mardukbp committed:
Customize the prompt

The prompt now uses the correct spelling SymPy

mardukbp commented:
> By the way, a minor issue: in the plugin, the input prompt says "Sympy]", but it should be spelled "SymPy]".

Yeah, that is auto-generated by TeXmacs. I have fixed that now.

asmeurer commented:
The tm_sympy file is already executable (this is tracked by git).

mardukbp commented:
I don't think so. When Python cannot find SymPy, the exception is printed in TeXmacs. I checked the source code of TeXmacs, the problem seems to be that Qt cannot launch the process. Is there any character that needs quoting or a non-ASCII character in the path of `tm_sympy`?

I will try to take a closer look in the coming days.

asmeurer commented:
No. There aren't any spaces in the path either.

oscarbenjamin commented:
What's the status of this?

asmeurer commented:
I wasn't able to get it working, though that could just be because I am using macOS. If someone can verify that it works on Linux, we can merge this and fix the Mac issues later.

jksuom commented:
I am testing on Ubuntu 14.04. I had some trouble adding SymPy to the search path but, after I managed to do that, the welcome message appeared all right. What kind of tests should be done in addition?

czgdp1807 commented:
@asmeurer @jksuom Can it be merged? The last few comments show that it is ready to go in.

czgdp1807 commented:
@asmeurer @oscarbenjamin @jksuom Is this ready for merge?
